src/framework/utils/quakemdl.zig

```diff
@@ -0,0 +1,737 @@
+const std = @import("std");
+
+const colors = @import("../colors.zig");
+const debug = @import("../debug.zig");
+const default_mesh = @import("../graphics/shaders/default-mesh.glsl.zig");
+const images = @import("../images.zig");
+const math = @import("../math.zig");
+const mem = @import("../mem.zig");
+const mesh = @import("../graphics/mesh.zig");
+const graphics = @import("../platform/graphics.zig");
+
+const assert = std.debug.assert;
+const Allocator = std.mem.Allocator;
+const File = std.fs.File;
+
+pub const MDL = struct {
+    frames: []MDLFrameType,
+    skins: []MDLSkinType
+};
+
+pub const MDLFrameTypeTag = enum {
+    single,
+    group
+};
+
+pub const MDLFrameType = union(MDLFrameTypeTag) {
+    single: MDLFrame,
+    group: MDLFrameGroup
+};
+
+pub const MDLFrame = struct {
+    name: [16]u8,
+    mesh: mesh.Mesh,
+};
+
+pub const MDLFrameGroup = struct {
+    meshes: []mesh.Mesh,
+    intervals: []f32
+};
+
+pub const MDLSkinTypeTag = enum {
+    single,
+    group
+};
+
+pub const MDLSkinType = union(MDLSkinTypeTag) {
+    single: MDLSkin,
+    group: MDLSkinGroup
+};
+
+pub const MDLSkin = struct {
+    texture: graphics.Texture,
+};
+
+pub const MDLSkinGroup = struct {
+    textures: []graphics.Texture,
+    intervals: []f32
+};
+
+const MDLFileHeader_ = extern struct {
+    magic: [4]u8,
+    version: u32,
+    scale: [3]f32,
+    origin: [3]f32,
+    radius: f32,
+    offsets: [3]f32,
+    skin_count: u32,
+    skin_width: u32,
+    skin_height: u32,
+    vertex_count: u32,
+    triangle_count: u32,
+    frame_count: u32,
+    sync_type: u32,
+    flags: u32,
+    size: u32,
+
+    pub fn read(file: File) !MDLFileHeader_ {
+        const size = @sizeOf(MDLFileHeader_);
+        var bytes: [size]u8 = undefined;
+        _ = try file.read(&bytes);
+
+        return @bitCast(bytes);
+    }
+};
+
+const MDLMeshBuildConfig_ = struct {
+    stvertexes: []STVertex_,
+    triangles: []Triangle_,
+    skin_width: f32,
+    skin_height: f32,
+    transform: math.Mat4,
+    material: graphics.Material
+};
+
+const MDLSkin_ = struct {
+    type: u32,
+    width: u32,
+    height: u32,
+    pixels: []u8,
+
+    pub fn read(allocator: Allocator, file: File, width: u32, height: u32) !MDLSkin_ {
+        // Skin type
+        var bytes: [4]u8 = undefined;
+        _ = try file.read(&bytes);
+        const skin_type: u32 = @bitCast(bytes);
+        assert(skin_type == 0);
+
+        // Skin pixels
+        const size: u32 = width * height;
+        const pixels: []u8 = try allocator.alloc(u8, size);
+        _ = try file.read(pixels);
+        //defer allocator.free(indexes);
```

Comment:
Memory management in Zig is the area where I need the most help.

Reply:
In Zig, things that are allocated like this generally keep track of their allocator and implement a `deinit` function that can be called to free their memory later